src/bun.js/api/bun/h2_frame_parser.zig

```diff
@@ -0,0 +1,2440 @@
+        result.put(globalObject, JSC.ZigString.static("inflateDynamicTableSize"), JSC.JSValue.jsNumber(settings.headerTableSize));
+
+        // TODO: make this  real?
+        result.put(globalObject, JSC.ZigString.static("outboundQueueSize"), JSC.JSValue.jsNumber(0));
+        return result;
+    }
+    pub fn goaway(this: *H2FrameParser, globalObject: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) callconv(.C) JSValue {
+        JSC.markBinding(@src());
+        const args_list = callframe.arguments(3);
+        if (args_list.len < 1) {
+            globalObject.throw("Expected errorCode argument", .{});
+            return .zero;
+        }
+
+        const error_code_arg = args_list.ptr[0];
+
+        if (!error_code_arg.isNumber()) {
+            globalObject.throw("Expected errorCode to be a number", .{});
+            return .zero;
+        }
+        const errorCode = error_code_arg.toInt32();
+        if (errorCode < 1 and errorCode > 13) {
+            globalObject.throw("invalid errorCode", .{});
+        }
+
+        var lastStreamID = this.lastStreamID;
+        if (args_list.len >= 2) {
+            const last_stream_arg = args_list.ptr[1];
+            if (!last_stream_arg.isEmptyOrUndefinedOrNull()) {
+                if (!last_stream_arg.isNumber()) {
+                    globalObject.throw("Expected lastStreamId to be a number", .{});
+                    return .zero;
+                }
+                const id = last_stream_arg.toInt32();
+                if (id < 0 and id > MAX_STREAM_ID) {
+                    globalObject.throw("Expected lastStreamId to be a number between 1 and 2147483647", .{});
+                    return .zero;
+                }
+                lastStreamID = @intCast(id);
+            }
+            if (args_list.len >= 3) {
+                const opaque_data_arg = args_list.ptr[2];
+                if (!opaque_data_arg.isEmptyOrUndefinedOrNull()) {
+                    if (opaque_data_arg.asArrayBuffer(globalObject)) |array_buffer| {
+                        var slice = array_buffer.slice();
+                        this.sendGoAway(0, @enumFromInt(errorCode), slice, lastStreamID);
+                        return JSC.JSValue.jsUndefined();
+                    }
+                }
+            }
+        }
+
+        this.sendGoAway(0, @enumFromInt(errorCode), "", lastStreamID);
+        return JSC.JSValue.jsUndefined();
+    }
+
+    pub fn ping(this: *H2FrameParser, globalObject: *JSC.JSGlobalObject, callframe: *JSC.CallFrame) callconv(.C) JSValue {
+        JSC.markBinding(@src());
+        const args_list = callframe.arguments(1);
+        if (args_list.len < 1) {
+            globalObject.throw("Expected payload argument", .{});
+            return .zero;
+        }
+
+        if (args_list.ptr[0].asArrayBuffer(globalObject)) |array_buffer| {
+            var slice = array_buffer.slice();
+            this.sendPing(false, slice);
```

Comment:
sendPing happens immediately or otherwise clones the data to be buffered right?

Reply:
yeah